lib/protobuf/message.rb

```diff
@@ -140,20 +140,54 @@ def to_proto
     def ==(other)
       return false unless other.is_a?(self.class)
       each_field do |field, value|
-        return false unless value == other.__send__(field.name)
+        return false unless value == other[field.name]
       end
       true
     end
 
     def [](name)
       if (field = self.class.get_field(name, true))
-        __send__(field.getter)
+        if field.repeated?
+          @values[field.name] ||= ::Protobuf::Field::FieldArray.new(field)
+        elsif @values.key?(field.name)
+          @values[field.name]
+        else
+          field.default_value
+        end
+      else
+        fail ArgumentError, "invalid field name=#{name.inspect}"
       end
     end
 
     def []=(name, value)
       if (field = self.class.get_field(name, true))
-        __send__(field.setter, value) unless value.nil?
+        @encode = nil
+        if field.repeated?
+          if value.is_a?(Array)
+            value = value.dup
+            value.compact!
```

Comment:
I know this is preserving existing behavior, but just want to note that IMO this `compact` thing is super weird behavior, settings an array with nil values in there should def be an error